### DISCUSSION TITLE
Counter example output

### DISCUSSION THREAD

**Discussion by ghmanoj** [JakeWharton/mosaic#36] [Q&A] [answered]
Does the counter example require specific terminal to run? ( I am using zsh shell )
I followed the example given but output is in new line at each counter update.

<img width="406" alt="Screen Shot 2021-06-25 at 9 56 47 PM" src="https://user-images.githubusercontent.com/10175663/123502303-3e63bd00-d600-11eb-9eee-9b9e0a4f398a.png">

**Comment by ghmanoj** [comment, marked as answer]
Ok figured it out. The ./gradlew run was messing up the terminal output.
After I created the distribution jar and ran from it it worked.

<img width="331" alt="Screen Shot 2021-06-25 at 10 02 52 PM" src="https://user-images.githubusercontent.com/10175663/123502422-188ae800-d601-11eb-981b-4ecebff118d6.png">

**Reply by JakeWharton** [reply]
Yeah it seems Gradle strips the ANSI control characters which allow repositioning the cursor and erasing lines. Glad you figured it out.

**Reply by ColtonIdle** [reply]
@JakeWharton is there any "easy" way to develop without having to create a jar every time to test my code? I'm not really a java cmd line dev and so I might be missing something obvious. My first two goto is to run in intellij and second is to run via gradle. Building and testing via the jar seems fairly cumbersome to me?

**Reply by JakeWharton** [reply]
You could file a feature request on IntelliJ to have an option to honor ANSI control codes. Otherwise no, I don't think there's any other way. A terminal which honors the codes is required.

**Reply by ColtonIdle** [reply]
Alright. I will try to automate the jar creation so I can test on a fast feedback loop. Unchartered waters for me, so hopefully this is simple. =)

**Reply by ColtonIdle** [reply]
Alright. So trying to create a jar keeps failing with 

```
Kotlin: [Internal Error] java.lang.IllegalStateException: The provided plugin androidx.compose.compiler.plugins.kotlin.ComposeComponentRegistrar is not compatible with this version of compiler
```

@ghmanoj any tips on how you created a jar (if you used intellij)?

Going off of Jakes rec to file a feature request on intellij, I decided to just try to run ./gradlew run in my actual terminal and it gave me the same output. So is the issue intellij or gradle (or both?)

**Reply by JakeWharton** [reply]
Both IntelliJ's terminal and Gradle's output will remove ANSI control characters. They are separate, but perform the same action which prevents Mosaic (or anything using ANSI characters) from working.

**Reply by DavidEdwards** [reply]
I also cannot create jars. This seems to be an issue with the kotlin version (mismatch with IDE and project).

In the meantime, I have taken to building the project in the IDE, and running it once to get the full run command.

![image](https://user-images.githubusercontent.com/1585923/123990482-19899400-d9ca-11eb-9623-c88f3a2678af.png)

You can use that in a standard terminal verbatim.

**Reply by ghmanoj** [reply]
@ColtonIdle 
I used application plugin for gradle, set main class name then run distZip gradle task. That worked fine for me.

**Reply by ColtonIdle** [reply]
@ghmanoj thank you! I was able to run the app by running ./gradlew distZip then navigating to build > distributions > myname.zip then unzip, then double click on bin > myname

Thankfully this works, but I am still very interested to see if anyone has a practical way to develo/run code with mosaic.

**Reply by JakeWharton** [reply]
Run installDist, not distZip.

See https://github.com/JakeWharton/mosaic#samples

**Reply by ColtonIdle** [reply]
Aha. Okay. installDist does indeed make this easier. Sorry I wasn't using a sample I went through the wizard myself which is how I missed that part of the docs!

Awesome. It works! Now to start building! I'm going to attempt to replace some bash scripts with mosaic. Let's see what happens. 😄

**Reply by sschuberth** [reply]
> Both IntelliJ's terminal and Gradle's output will remove ANSI control characters.

For the record, I don't believe these are actively removing ANSI control characters. It's more like whatever underlying logic detects the terminal's capabilities fails when output is redirected (as done with Gradle's `run` task from the `application` plugin) and falls back to plain text output even before Gradle sees it.

**Reply by ColtonIdle** [reply]
@sschuberth you think we should file a bug on youtrack? I'd love to improve the dev cycle here.

**Reply by sschuberth** [reply]
I don't think the problem is solvable, so no. Once output is redirected (which it needs to be with `./gradlew run`), there is no way for the actual program to detect the actual terminal's capabilities (as redirection could also happen to a file).

**Reply by JakeWharton** [reply]
> I don't believe these are actively removing ANSI control characters. It's more like whatever underlying logic detects the terminal's capabilities fails when output is redirected

Mosaic does not use terminal detection. Both Gradle and IntelliJ strip ANSI control sequences.

You can try it yourself with
```kotlin
fun main() {
  for (i in 0..20) {
    println("The count is $i")
    Thread.sleep(1_000)
    print("\u001B[F\u001B[K")
  }
}
```

**Reply by ColtonIdle** [reply]
@JakeWharton do you think it's fair to file a bug on intellij/youtrack for this?

**Reply by JakeWharton** [reply]
IntelliJ could support it. I'm not sure why they don't. Maybe it's just not highly-sought after. Colors do work, though.

Gradle is a bit more challenging because the output gets wrapped into their output which is already using ANSI control sequences to do complex rewriting of the output. So they would have to either wash their hands of responsibility and allow the program to have the ability to write over the existing Gradle output, or they would have to maintain the cursor position of the program themselves and somehow clip it so that it doesn't go over their output. I'm not sure what they'd say.

You can certainly try. I'd be interested to see what their takes on it is.
